File: haskell/2015/day19b/src/Main.hs
module Main (main) where

import Data.Int (Int32)
import System.Environment (getArgs, getProgName)
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)
import Text.Parsec
import Text.Parsec.String
import Control.DeepSeq
import System.Clock

usage :: String -> IO ()
usage progname = do
    hPutStrLn stderr $ "usage: " ++ progname ++ " <input file>"
    exitFailure

element :: Parser String
element = do
          first <- upper
          rest  <- many lower
          return (first : rest)

molecule :: Parser [String]
molecule = many1 element <* eof

process :: String -> Int32
process content =
    let moleculeLine = last (lines content)
    in case parse molecule "" moleculeLine of
           Left err -> error (show err)
           Right elts -> fromIntegral $ countSteps elts
  where
    countSteps :: [String] -> Int
    countSteps elts =
        let n    = length elts
            rnAr = length [t | t <- elts, t == "Rn" || t == "Ar"]
            ys   = length [t | t <- elts, t == "Y"]
        in n - rnAr - 2 * ys - 1


showTime :: TimeSpec -> String
showTime elapsed =
    let ns = fromIntegral (toNanoSecs elapsed) :: Double
    in if ns < 1000
       then show ns ++ " ns"
       else if ns < 1000000
       then show (ns / 1000.0) ++ " μs"
       else if ns < 1000000000
            then show (ns / 1000000.0) ++ " ms"
            else show (ns / 1000000000.0) ++ " s"
main :: IO ()
main = do
    args <- getArgs
    progname <- getProgName
    case args of
        [filename] -> do
            start <- getTime Monotonic
            content <- readFile filename
            let result = process content
            result `deepseq` return ()
            end <- getTime Monotonic
            let elapsed = diffTimeSpec start end
            putStrLn $ "result = " ++ show result
            putStrLn $ "elapsed time: " ++ showTime elapsed
        _ -> usage progname

File: rust/2015/day19b/src/main.rs
use std::env;
use std::fs;
use std::io;
use std::process;

use regex::Regex;

type Element = String;

fn usage(progname: &str) {
    eprintln!("usage: {progname} <input file>");
    process::exit(1);
}

fn parse_file(filename: &str) -> io::Result<Vec<Element>> {
    let content = fs::read_to_string(filename)?;
    let last_line = content.lines().last().unwrap();
    let re = Regex::new(r"([A-Z][a-z]*)").unwrap();
    let mut elements: Vec<Element> = vec![];
    for (_, [element]) in re.captures_iter(last_line).map(|c| c.extract()) {
        elements.push(element.to_string());
    }
    Ok(elements)
}

fn process_file(filename: &str) -> io::Result<u32> {
    let elements = parse_file(filename)?;
    let result = elements.len()
        - elements
            .iter()
            .filter(|x| x.as_str() == "Rn" || x.as_str() == "Ar")
            .count()
        - (2 * elements.iter().filter(|x| x.as_str() == "Y").count())
        - 1;
    Ok(result as u32)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let progname = &args[0];
    if args.len() < 2 {
        usage(progname);
    }
    let filename = &args[1];
    match process_file(filename) {
        Ok(result) => println!("result = {result}"),
        Err(e) => {
            eprintln!("error while processing file `{filename}': {e}");
            process::exit(1);
        }
    }
}
